src/windows/ScrollView.tsx

```diff
@@ -9,28 +9,31 @@
 
 import React = require('react');
 import RN = require('react-native');
-import { ScrollView as ScrollViewBase } from '../native-common/ScrollView';
 
+import { ScrollView as ScrollViewBase } from '../native-common/ScrollView';
 import EventHelpers from '../native-common/utils/EventHelpers';
+import Types = require('../common/Types');
 
 export class ScrollView extends ScrollViewBase {
 
-    protected _render(props: RN.ScrollViewProps): JSX.Element {
+    protected _render(props: Types.ScrollViewProps): JSX.Element {
         var onKeyDownCallback = this.props.onKeyPress ? this._onKeyDown : undefined;
 
         // TODO: #737970 Remove special case for UWP when this bug is fixed. The bug
         //   causes you to have to click twice instead of once on some pieces of UI in
         //   order for the UI to acknowledge your interaction.
         const keyboardShouldPersistTaps = 'always';
 
+        let nativeProps = {
+            ...props,
+            onKeyDown: onKeyDownCallback,
+            keyboardShouldPersistTaps: keyboardShouldPersistTaps,
+            tabNavigation: this.props.tabNavigation,
+            disableKeyboardBasedScrolling: true
+    } as RN.ScrollViewProps;
```

Comment:
> as RN.ScrollViewProps [](start = 5, length = 22)

Is this tradeoff worth it?

Reply:
Are you talking about this particular case, or is this a broader question? 

In any case, I think the answer is yes. We've been maintaining this Frankenstein "react-native.d.ts" file for some time now, and it's getting pretty unmaintainable. The use of the public one requires a few more ugly casts in our code, but I still think it's worth it.

Reply:
It was a broader Q.

I was thinking whether we can find a better way to describe the new "extended" properties, currently they look like magic (for ex. we use "tabNavigation" because we "know" RNW supports that, but it's not documented anywhere inside ReactXP).
I made an experiment:

1. I added this to the extension file:

    interface ExtendedScrollViewProps extends RN.ScrollViewProps {
         // Windows only
         onKeyDown? : Function;
         onKeyUp? : Function;
         tabNavigation? : 'local' | 'cycle' | 'once';
         disableKeyboardBasedScrolling?: boolean;       
    }

2. Converted the code to:

        let nativeProps: RN.ExtendedScrollViewProps = {
            ...props as RN.ScrollViewProps, // unrelated hack, forces a Types.ScrollViewProps into RN.ScrollViewProps
            onKeyDown: onKeyDownCallback,
            keyboardShouldPersistTaps: keyboardShouldPersistTaps,
            tabNavigation: this.props.tabNavigation,
            disableKeyboardBasedScrolling: true
    };

It seems to work, at least in this case.


---
In reply to: [194625530](https://github.com/Microsoft/reactxp/pull/677#discussion_r194625530) [](ancestors = 194625530)